File: dbflow/src/main/java/com/raizlabs/android/dbflow/sql/language/Delete.java
package com.raizlabs.android.dbflow.sql.language;

import com.raizlabs.android.dbflow.sql.Query;
import com.raizlabs.android.dbflow.sql.QueryBuilder;
import com.raizlabs.android.dbflow.structure.Model;

/**
 * Description: Constructs the beginning of a SQL DELETE query
 */
public class Delete implements Query {

    /**
     * Deletes the specified table
     *
     * @param table        The table to delete
     * @param conditions   The list of conditions to use to delete from the specified table
     * @param <TModel> The class that implements {@link com.raizlabs.android.dbflow.structure.Model}
     */
    public static <TModel extends Model> void table(Class<TModel> table, SQLCondition... conditions) {
        new Delete().from(table).where(conditions).query();
    }

    /**
     * Deletes the list of tables specified.
     * WARNING: this will completely clear all rows from each table.
     *
     * @param tables The list of tables to wipe.
     */
    @SafeVarargs
    public static void tables(Class<? extends Model>... tables) {
        for (Class modelClass : tables) {
            table(modelClass);
        }
    }

    /**
     * Returns the new SQL FROM statement wrapper
     *
     * @param table        The table we want to run this query from
     * @param <TModel> The table class
     * @return
     */
    public <TModel extends Model> From<TModel> from(Class<TModel> table) {
        return new From<>(this, table);
    }

    @Override
    public String getQuery() {
        return new QueryBuilder()
                .append("DELETE")
                .appendSpace().getQuery();
    }
}

File: dbflow/src/main/java/com/raizlabs/android/dbflow/sql/language/Delete.kt
package com.raizlabs.android.dbflow.sql.language

import com.raizlabs.android.dbflow.sql.Query

/**
 * Description: Constructs the beginning of a SQL DELETE query
 */
class Delete : Query {

    override val query: String
        get() = "DELETE "

    /**
     * Returns the new SQL FROM statement wrapper
     *
     * @param table    The table we want to run this query from
     * @param [T] The table class
     * @return [T]
     **/
    fun <T : Any> from(table: Class<T>): From<T> = From(this, table)

    companion object {

        /**
         * Deletes the specified table
         *
         * @param table      The table to delete
         * @param conditions The list of conditions to use to delete from the specified table
         * @param [T]   The class that implements [com.raizlabs.android.dbflow.structure.Model]
         */
        fun <T : Any> table(table: Class<T>, vararg conditions: SQLOperator) {
            Delete().from(table).where(*conditions).executeUpdateDelete()
        }

        /**
         * Deletes the list of tables specified.
         * WARNING: this will completely clear all rows from each table.
         *
         * @param tables The list of tables to wipe.
         */
        fun tables(vararg tables: Class<*>) {
            tables.forEach { table(it) }
        }
    }
}


inline fun <reified T : Any> delete() = SQLite.delete(T::class.java)

inline fun <reified T : Any> delete(deleteClause: From<T>.() -> BaseModelQueriable<T>)
        = deleteClause(SQLite.delete(T::class.java))